tracer/src/Datadog.Trace/DataStreamsMonitoring/DataStreamsWriter.cs

```diff
@@ -0,0 +1,175 @@
+﻿// <copyright file="DataStreamsWriter.cs" company="Datadog">
+// Unless explicitly stated otherwise all files in this repository are licensed under the Apache 2 License.
+// This product includes software developed at Datadog (https://www.datadoghq.com/). Copyright 2017 Datadog, Inc.
+// </copyright>
+
+#nullable enable
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+using Datadog.Trace.DataStreamsMonitoring.Aggregation;
+using Datadog.Trace.DataStreamsMonitoring.Transport;
+using Datadog.Trace.ExtensionMethods;
+using Datadog.Trace.Logging;
+using Datadog.Trace.Util;
+
+namespace Datadog.Trace.DataStreamsMonitoring;
+
+internal class DataStreamsWriter
+{
+    private static readonly IDatadogLogger Log = DatadogLogging.GetLoggerFor<DataStreamsWriter>();
+
+    private readonly BoundedConcurrentQueue<StatsPoint> _buffer = new(queueLimit: 10_000);
+    private readonly Task _processTask;
+    private readonly ManualResetEventSlim _processingMutex = new(initialState: false, spinCount: 0);
+    private readonly TaskCompletionSource<bool> _processExit = new(TaskCreationOptions.RunContinuationsAsynchronously);
+    private readonly DataStreamsAggregator _aggregator;
+    private readonly IDataStreamsApi _api;
+    private readonly Timer _flushTimer;
+    private byte[]? _serializationBuffer;
+    private long _pointsDropped;
+    private int _flushRequested;
+
+    public DataStreamsWriter(
+        DataStreamsAggregator aggregator,
+        IDataStreamsApi api,
+        long bucketDurationMs)
+    {
+        _aggregator = aggregator;
+        _api = api;
+
+        _processTask = Task.Run(ProcessQueueLoopAsync);
+        _processTask.ContinueWith(t => Log.Error(t.Exception, "Error in processing task"), TaskContinuationOptions.OnlyOnFaulted);
+
+        _flushTimer = new Timer(
+            x => ((DataStreamsWriter)x!).RequestFlush(),
+            this,
+            dueTime: bucketDurationMs,
+            period: bucketDurationMs);
+    }
+
+    public void Add(in StatsPoint point)
+    {
+        if (_buffer.TryEnqueue(point))
+        {
+            if (!_processingMutex.IsSet)
```

Comment:
This looks like it will request a flush every time a point is added, is that the desired behavior?